scanner/src/main/kotlin/storages/Sw360Storage.kt

```diff
@@ -0,0 +1,143 @@
+
+/**
+ * The SW360 storage back-end uses the sw360-client library in order to read/add attachments from the configured
+ * SW360 instance.
+ */
+class Sw360Storage(
+    sw360Configuration: Sw360StorageConfiguration
+) : ScanResultsStorage() {
+    private val sw360JsonMapper = jsonMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+    private val sw360ConnectionFactory = createSw360Connection(sw360Configuration, sw360JsonMapper)
+    private val releaseClient = sw360ConnectionFactory.releaseAdapter
+
+    override fun readFromStorage(id: Identifier): Result<ScanResultContainer> =
+        try {
+            val scanResults = releaseClient.getSparseReleaseByNameAndVersion(createReleaseName(id), id.version)
+                .flatMap { releaseClient.getReleaseById(it.releaseId) }
+                .map { getScanResultOfRelease(it) }
+                .orElse(emptyList())
+                .map { path ->
+                    yamlMapper.readValue<ScanResult>(path.toFile())
+                }
+            Success(ScanResultContainer(id, scanResults))
+        } catch (e: SW360ClientException) {
+            val message = "Could not read scan results for ${id.toCoordinates()} in SW360."
+            log.info { message }
+            Failure(message)
+        }
+
+    override fun addToStorage(id: Identifier, scanResult: ScanResult): Result<Unit> =
+        try {
+            val cachedScanResult = tmpStorageFile(id)
+            yamlMapper.writeValue(cachedScanResult, scanResult)
+
+            val uploadResult = releaseClient.getSparseReleaseByNameAndVersion(createReleaseName(id), id.version)
+                .flatMap { releaseClient.getReleaseById(it.releaseId) }
+                .map { createAttachmentOfScanResult(it, cachedScanResult.toPath()) }
+                .map { releaseClient.uploadAttachments(it) }
+
+            if (uploadResult.isPresent && uploadResult.get().isSuccess) {
+                log.debug { "Stored scan result for '${id.toCoordinates()}' in SW360." }
+                Success(Unit)
+            } else {
+                Failure("Failed to add scan results for '${id.toCoordinates()}' in SW360.")
+            }
+        } catch (e: SW360ClientException) {
+            val message = "Failed to add scan results for '${id.toCoordinates()}' in SW360: " +
+                    e.collectMessagesAsString()
+            log.info { message }
+            Failure(message)
+        }
+
+    private fun createReleaseName(id: Identifier): String =
+        listOfNotNull(id.namespace, id.name).joinToString("/")
+
+    private fun getScanResultOfRelease(release: SW360Release): List<Path> =
+        release.embedded.attachments
+            .filter {
+                it.attachmentType == SW360AttachmentType.SCAN_RESULT_REPORT && it.filename == SCAN_RESULTS_FILE_NAME
+            }
+            .mapNotNull { releaseClient.downloadAttachment(release, it, createTempDirectory()).orElse(null) }
```

Comment:
Both, the temporary directory and the downloaded attachment file need to be deleted again after processing. Otherwise, the temp directory will be filled with garbarge.